pkg/metadata/provider_azure_imds.go

```diff
@@ -0,0 +1,193 @@
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path"
+	"strings"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+)
+
+// ProviderAzureIMDS reads from Azure's Instance Metadata Service (IMDS) API.
+type ProviderAzureIMDS struct {
+	client      *http.Client
+	providerOVF *ProviderAzureOVF
+}
+
+// NewAzureIMDS factory
+func NewAzureIMDS() *ProviderAzureIMDS {
+	client := &http.Client{
+		Timeout: time.Second * 2,
+	}
+	return &ProviderAzureIMDS{
+		client:      client,
+		providerOVF: NewAzureOVF(client),
+	}
+}
+
+func (p *ProviderAzureIMDS) String() string {
+	return "Azure-IMDS"
+}
+
+// Probe checks if Azure IMDS API is available
+func (p *ProviderAzureIMDS) Probe() bool {
+	// "Poll" VM Unique ID
+	// See: https://azure.microsoft.com/en-us/blog/accessing-and-using-azure-vm-unique-id/
+	pollVMID := func() error {
+		_, err := p.imdsGet("compute/vmId")
+		return err
+	}
+	if err := retry(6, 5*time.Second, pollVMID); err != nil {
```

Comment:
As far as I can see no other `Probe` function does a retry. Is this really necessary? If all providers would probe like this it potentially will take a long time to cycle through the list of all providers until we hit a match.

Reply:
You're right, for the IMDS provider it's probably not needed, I'll have time to test this tomorrow. For the OVF provider, as mentioned earlier, it is needed and I pretty much mimmick what WALinuxAgent does. I'll remove the retry logic for the IMDS provider and move the retry function to `provider_azure_ovf.go`.

Reply:
Ah, I missed your earlier comment. Can we remove the OVF provider from the default list? I suspect a common usage for the `metadata` package would be just add it without arguments. Having it taking 30s longer depending on which provider is used and where in the list the provider happens to be does not look like a good user experience. `Probe()` is really meant to be a quick check. Is there a different way to figure out which provider to use?

Reply:
@rn see my latest comment, please.

* I removed the OVF provider from the default list
* I moved the Azure provider to the end of the provider list which is the best we can do IMO.

This way using `metadata` without args, the 30 second delay never occurs _if running on a supported platform_.

However, when using the `metadata` package without args, in an environment where none of the providers are supported, a 30 second delay will occur. I don't know if this is an actual use case, though. If it is, we'd have to _always_ exclude the `azure` provider by default and in case one wanted to use it, `explicitly include it` which is not a nice user experience, too, IMO.